src/internal/grpcutil/iter.go

```diff
@@ -0,0 +1,37 @@
+package grpcutil
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+	"google.golang.org/grpc"
+)
+
+type ClientStream[T any] interface {
+	Recv() (*T, error)
+	grpc.ClientStream
+}
+
+type Iterator[T any] struct {
+	cs ClientStream[T]
+}
+
+func NewIterator[T any](cs ClientStream[T]) Iterator[T] {
+	return Iterator[T]{cs: cs}
+}
+
+func (it Iterator[T]) Next(ctx context.Context, x *T) error {
+	return it.cs.RecvMsg(x)
+}
+
+func ForEach[T any](cs ClientStream[T], fn func(x T) error) error {
```

Comment:
Yeah, this is why I think the constraint should be a pointer; fn should take a *T if Recv returns *T.

Reply:
That seams reasonable to me.  The big thing with `ForEach` is that the closure can't retain the object.  I guess it doesn't matter if it gets passed a pointer or not?